engine/src/execution_graph/logic_controllers/BatchProcessing.cpp

```diff
@@ -278,6 +289,11 @@ kstatus BindableTableScan::run() {
             //this will allow us to prevent from having too many open file handles by being
             //able to limit the number of file tasks
             auto handle = provider->get_next(true);
+            if (provider->is_sql()) {
+              if (handle.sql_handle.row_count == 0) {
```

Comment:
This should be handled by the `has_next()` function that was just called. If it returned true, then the row count here should never be 0. Is there a reason we are doing this?

Reply:
Ok, I looked at the code and understand why you need to do this. What I would suggest is that `mysql_data_provider` be able to always hold a result set. For example:
It could have two extra members:
```
data_handle current_handle;
bool have_current_handle;  // initialized to false
```
then
```
bool mysql_data_provider::has_next() {
    if (!this->table_fetch_completed  && !have_current_handle){
         fetch_data();
   }
  return (this->table_fetch_completed == false);
}
```
`fetch_data()` would have the logic that is in `get_next()` that obtains and builds the data_handle.
`have_current_handle` would be set to true by `fetch_data` and set to false by `get_next`.

You can then take it to the next step and make it so that when you call `get_next()` after it returns the handle, it runs `fetch_data` on a separate thread so that it is already fetching the next batch while the rest of the engine does parsing or anything else. This async behaviour could be configurable

Reply:
@williamBlazing It seems the `async` optimization would be a nice to have, I've already resolved the main issue regarding to the `is_sql`, the use of `next` and the use of `CacheDataIO`. How about we implement async data_providers in a next PR? cc @felipeblazing @gcca